File: lana/src/main/scala/com/financialforce/lana/display/QuickPickWorkspace.scala
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */
package com.financialforce.lana.display

import com.financialforce.lana.Context
import com.financialforce.lana.runtime.vscode.window
import io.scalajs.nodejs.path.Path

import scala.concurrent.Future
import scala.concurrent.Future.{failed, successful}
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue

object QuickPickWorkspace {

  def pickOrReturn(context: Context): Future[String] =
    if (context.workspaces.length > 1) {
      QuickPick.pick(context.workspaces map { ws =>
        new QuickPick.Item(ws.name, ws.path, "")
      }, new QuickPick.Options("Select a workspace:")) flatMap {
        case results if results.length == 1 => successful(results.head.description)
        case _ => failed(new Exception("No workspace selected"))
      }
    } else if (context.workspaces.length == 1) {
      successful(context.workspaces.head.path)
    } else {
      val parent = window.activeTextEditor.flatMap(te => Path.parse(te.document.uri.fsPath).dir)
      parent.map(p => successful(p)).getOrElse(failed(new Exception("No workspace selected")))
    }

}

File: lana/src/display/QuickPickWorkspace.ts
/*
 * Copyright (c) 2020 Certinia Inc. All rights reserved.
 */
import { window } from 'vscode';
import { Utils } from 'vscode-uri';

import type { Context } from '../Context.js';
import { VSWorkspace } from '../workspace/VSWorkspace.js';
import { Item, Options, QuickPick } from './QuickPick.js';

export class QuickPickWorkspace {
  static async pickOrReturn(context: Context): Promise<VSWorkspace> {
    const workspaceFolders = context.workspaceManager.workspaceFolders;

    if (workspaceFolders.length > 1) {
      const [workspace] = await QuickPick.pick(
        workspaceFolders.map((ws) => new Item(ws.name(), ws.uri, '')),
        new Options('Select a workspace:'),
      );

      if (workspace) {
        const selectedWorkspace = workspaceFolders.find((ws) => ws.uri === workspace.description);
        if (!selectedWorkspace) {
          throw new Error('Selected workspace not found');
        }
        return selectedWorkspace;
      } else {
        throw new Error('No workspace selected');
      }
    } else if (workspaceFolders.length === 1) {
      const selectedWorkspace = workspaceFolders[0];
      if (!selectedWorkspace) {
        throw new Error('No workspace available');
      }
      return selectedWorkspace;
    } else {
      if (window.activeTextEditor) {
        const documentUri = window.activeTextEditor.document.uri;
        const folderUri = Utils.dirname(documentUri);
        return new VSWorkspace({
          uri: folderUri,
          name: Utils.basename(folderUri),
          index: 0,
        });
      } else {
        throw new Error('No workspace selected');
      }
    }
  }
}
